modules/dnn/src/cuda4dnn/csl/memory.hpp

```diff
@@ -0,0 +1,273 @@
+// of this distribution and at http://opencv.org/license.html.
+
+#ifndef OPENCV_DNN_CUDA4DNN_CSL_MEMORY_HPP
+#define OPENCV_DNN_CUDA4DNN_CSL_MEMORY_HPP
+
+#include "error.hpp"
+#include "pointer.hpp"
+
+#include <cuda_runtime_api.h>
+
+#include <cstddef>
+#include <memory>
+#include <type_traits>
+
+namespace cv { namespace dnn { namespace cuda4dnn { namespace csl {
+
+    /* @brief smart device pointer with allocation/deallocation methods
+     *
+     * ManagedPtr is a smart shared device pointer which also handles memory allocation.
+     *
+     * The cv qualifications of T are ignored. It does not make sense to allocate immutable
+     * or volatile device memory.
+     */
+    template <class T>
+    class ManagedPtr {
+        static_assert(std::is_standard_layout<T>::value, "T must satisfy StandardLayoutType");
+
+    public:
+        using element_type = typename std::remove_cv<T>::type;
+
+        using pointer = DevicePtr<element_type>;
+        using const_pointer = DevicePtr<typename std::add_const<element_type>::type>;
+
+        using size_type = std::size_t;
+
+        ManagedPtr() noexcept : wrapped{ nullptr }, n{ 0 } { }
+        ManagedPtr(const ManagedPtr&) noexcept = default;
+        ManagedPtr(ManagedPtr&& other) noexcept
+            : wrapped{ std::move(other.wrapped) }, n{ other.n }
+        {
+            other.reset();
+        }
+
+        /** allocates device memory for \p count number of elements
+         *
+         * Exception Guarantee: Strong
+         */
+        ManagedPtr(std::size_t count) {
+            if (count <= 0) {
+                CV_Error(Error::StsBadArg, "number of elements is zero or negative");
+            }
+
+            void* temp = nullptr;
+            CUDA4DNN_CHECK_CUDA(cudaMalloc(&temp, count * sizeof(element_type)));
+
+            auto ptr = typename pointer::pointer(static_cast<element_type*>(temp));
+            wrapped.reset(ptr, [](element_type* ptr) {
+                if (ptr != nullptr) {
+                    /* contract violation for std::shared_ptr if cudaFree throws */
+                    CUDA4DNN_CHECK_CUDA(cudaFree(ptr));
+                }
+            });
+
+            n = count;
+        }
+
+        ManagedPtr& operator=(ManagedPtr&& other) noexcept {
+            wrapped = std::move(other.wrapped);
+            n = other.n;
+            other.reset();
+            return *this;
+        }
+
+        size_type size() const noexcept { return n; }
+
+        void reset() noexcept { wrapped.reset(); n = 0; }
+
+        /**
+         * deallocates any previously allocated memory and allocates device memory
+         * for \p count number of elements
+         *
+         * @note might optimize to avoid unnecessary reallocation
+         *
+         * Exception Guarantee: Strong
+         */
+        void reset(std::size_t count) {
+            /* avoid unncessary reallocations */
+            if (count <= n && (count > n / 2 || count * sizeof(T) < 256))
```

Comment:
Is this the check you intend?  It's saying, don't reallocate if we have enough memory and the requested size is greater than half the size we currently have or smaller than some threshold.  So we only reallocate if the count is some middle size, not too big and not too small?

Reply:
That was done to free memory if the new size is considerably smaller than the old size. It's intended to deal with cases like resizing a 1GB tensor to 1KB tensor. Half might have been overkill though.

Reply:
Right, that seems sensible. But if the count is small, say 1, it will never release it.

Reply:
I assume you meant the code to say `n * sizeof(T) < 256` rather than `count * sizeof(T) < 256`

Reply:
I am wondering if I need to maintain two variables: `capacity` and `size`. While the existing code isn't wrong, it doesn't help optimize cases where a 1GB tensor is allocated first, then 1KB and then 1GB again.

Reply:
People will do stuff like run images through a CNN, where each image is somewhat different in size.  So that would cause tensors to reallocate unless you do something like that you are doing here. For some reason, cudaMalloc is super slow. So you definitely want to avoid the reallocations with this kind of capacity/size thing. I would just do the simple thing of only reallocating when you absolutely have to.  It is very unusual for tensors to start out really huge and then get resized to something small and stay that way. So I would replace `if (count <= n && (count > n / 2 || count * sizeof(T) < 256))` with just `if (count <= n)`.

Reply:
I'm using OpenCV's CPU backend as the backbone (even IE, Vulkan and every other backend in OpenCV uses it). OpenCV resets the entire network instance when the input shape changes. In this process, all the previous memory is deallocated (even the backend memory) and then allocated again.

Hence, it's not possible to avoid the `cudaMalloc` calls.

It also performs a full reset when the user changes the outputs that have to be retained. So contrary to what my previous comment said, full reallocation will be performed even in this case.

It's a problem with OpenCV's backbone code. I could ask for clarification in the project plan issue I had opened.

Reply:
Yeah, that's definitely not good. Is there a reason why the CUDA stuff needs to use OpenCV's backbone like this?  What happens exactly? Does OpenCV's code call the destructors for all the layers and everything else in the network and construct a new network, reloading parameters from somewhere, every time tensor shapes change?

The performance penalty for calling `cudaMalloc` like this really is quite severe. For instance, in dlib's dnn code I had initially set it up to rellocate whenever sizes changed because my opinion was that users should pick a fixed size and run with it. But people complained to me that this was inconvenient or confusing.  People who didn't pick a fixed size and didn't understand what was happening would experience pretty bad performance. So I eventually switched the tensor's allocation code to use a capacity and size rule just like `std::vector`, and like the one I was suggesting be used here. 

If OpenCV doesn't avoid reallocation when sizes change we will be subjected to the same stream of user complaints. Which are pretty valid complaints.

Reply:
**Overview:**
OpenCV's public API provides few classes: `Layer` and `Net`. The `Layer` classes have methods like `supportBackend`, `forward`, `getMemoryShapes`, etc. Note that the `forward` method is intended only for OpenCV's backend. The `Net` class provides methods like `forward`, `forwardAsync`, `setInput`, etc. The `Net` implementation handles all user requests and forwards them to the appropriate backend if it thinks it needs to.

The `DNN_BACKEND_OPENCV` (OpenCV's backend) uses `cv::Mat` (and `cv::UMat` for OpenCL) for storing outputs, weights, etc. There are a few abstract classes (`BackendWrapper` and `BackendNode`) which enforce a common interface for all backends. `BackendWrapper` is a wrapper around backend memory which provides methods to mark memory as dirty, perform memory transfers, etc. `BackendNode` represents a computational node.

`BackendWrapper` has a special constructor which takes in another `BackendWrapper`. This constructor creates a `BackendWrapper` which uses the same memory as the given wrapper but with a different shape (and size).

The term OpenCV's backbone (my own term) is mainly the code in `dnn.cpp`. It always sets up the network to be capable of running on CPU using the OpenCV backend irrespective of what backend the end-user prefers. It contains a class called `LayerData` which has members equivalent to `vector<Mat> inputBlobs, outputBlobs, internalBlobs` and `vector<Ptr<BackendWrapper>> inputWrappers, outputWrapers, internalWrappers`. The blobs are the memory which OpenCV would use for computation on CPU. It has a class called `BlobManager` which tracks how layers are using blobs. If it thinks that memory can be safely reused, it does so.

After initializing the CPU stuff, for every blob, it creates a `BackendWrapper` which wraps the blob.  If a CPU blob can be reused, logically the corresponding backend wrapper can also be reused too. Hence, it uses the special constructor of the backend wrapper to create a wrapper around an existing backend memory.

The `Layer` class has methods like `initHalide`, `initInferenceEngine`, `initVulkan`, etc. which return a corresponding `cv::Ptr<BackendNode>`. The `Net` implementation calls a special forward method which initiates forward pass computation on the backend nodes.

The OpenCV backbone takes care of initialization which includes constructing `Layer` objects and allocating memory. It creates backend wrappers and nodes around CPU memory and layers. It sits in between user API and the backend API. That's a general overview of how the backbone works.

---

**What is the CUDA backend doing?**
All the backends currently make use of `BackendWrapper` and `BackendNode`. For the CUDA backend, I am using the `BackendWrapper` to wrap a `csl::Tensor` object. I am not using `BackendNode` as it makes having custom layers impossible.

The backend wrappers are initialized by the backbone code of which we have no control over (or maybe just a little bit).

---

**Details:**

`clear`: deallocates all CPU memory 
https://github.com/YashasSamaga/opencv/blob/cuda4dnn-csl/modules/dnn/src/dnn.cpp#L1230-L1253

`setupNet`: almost a NOP if the `netWasAllocated` is `true`; otherwise, it invokes `clear` and then `allocateLayers`
https://github.com/YashasSamaga/opencv/blob/master/modules/dnn/src/dnn.cpp#L1255-L1345

`allocateLayers`: clears all backend wrappers, adds fake references to input blob and for the blobs which the user requests to be retained and then invokes `allocateLayer` for every layer
https://github.com/YashasSamaga/opencv/blob/master/modules/dnn/src/dnn.cpp#L2368-L2416

`allocateLayer`: creates the blobs and the backend wrappers
https://github.com/YashasSamaga/opencv/blob/cuda4dnn-csl/modules/dnn/src/dnn.cpp#L1913-L2033

`Net::forward` calls `setupNet` (which does stuff if it wasn't already setup) and then starts the forward pass.
 https://github.com/YashasSamaga/opencv/blob/cuda4dnn-csl/modules/dnn/src/dnn.cpp#L2930

`Net::setInput` sets the input blob of the data layer
https://github.com/YashasSamaga/opencv/blob/cuda4dnn-csl/modules/dnn/src/dnn.cpp#L3141-L3181

The `setInput` has a variable called `oldShape` which is set to true if the previous input shape and the new input shape are the same. At the end of the function, `netWasAllocated` is set to `false` if `oldShape` is `false`. Hence, when the input shape changes, it causes a series of reallocations which we have no control over.

Reply:
> Is there a reason why the CUDA stuff needs to use OpenCV's backbone like this? 

Every backend is using the OpenCV backbone. I have made some tweaks at some places to accomodate the missing backend nodes for the CUDA backend. Messing with the backbone affects every other backend.

> What happens exactly? Does OpenCV's code call the destructors for all the layers and everything else in the network and construct a new network, reloading parameters from somewhere, every time tensor shapes change?

It does not construct a new network. I apologize for exaggarating in my comment before the previous one. It reallocates the input and output blobs.

The layer objects are left untouched. The parameters won't be reloaded.

Reply:
Hmm, darn, that sounds unfortunate.  You should post an issue asking about this.  Maybe there is some simple way to resolve this.  It sounds like there isn't though. If there isn't then don't worry about it for now.  In the longer term, all calls to `cudaMalloc` could be replaced with a custom memory manager with a global memory pool behind it. That's how a number of tools deal with the slowness of `cudaMalloc`. It's not my favorite approach though because it either requires the memory pool to do a big GPU allocation up front, which is annoying since it makes the software act like it owns the whole world.  There are other memory pooling strategies that allocate as they go, but they are sometimes less memory efficient, but probably would be fine here. But don't worry about this for now. If the right thing to do is to replace `cudaMalloc` that can always be done later.